### ISSUE TITLE
Investigate delivering the site over HTTPS

### ISSUE THREAD

**Issue by kirb** [kirb/LegacyUpdate#3]
Cloudflare [told me once](https://twitter.com/eastdakota/status/959952517395181568) that they worked out a solution to provide current, valid SHA1 certs for certain customers. I wonder if they would be willing to provide this kind of assistance to us as a FOSS community project? Of course, they'd also need to provide a dedicated IP because of the lack of SNI, so it's not exactly free for them.

If not, the next best thing would be to install a self-signed SHA1 cert. Not as tidy a solution, and it'd need to bypass Cloudflare, but it'll work. I'd need to rent a separate dedicated IP for this approach, again because of no SNI.

**Comment by kirb** [comment]
That’s an interesting approach, I like it, thanks for posting! Hopefully I should be able to proxy_pass to `https://legacyupdate.net:443`, with that name being routed to 127.0.0.1 in hosts. (There might be a way to tell proxy_pass which SNI to use though)

I heard back from Cloudflare, who gave me the bad news that they can’t provide SHA1 certs any more, because the root they used for this expired in June 2019. Realistically, because CF only supports using a customer-provided cert on Enterprise which is far too expensive ($200/month), it seems I’ll have to migrate to a strategy that looks something like:

* legacyupdate.net, run through Cloudflare, providing unencrypted http and modern TLS for clients with SNI (Vista+ and anyone just browsing the site on a modern device)
* sha1.legacyupdate.net, supporting only SSLv3/TLSv1.0 with either a Google cert if possible, or worst case, a self-signed root the Legacy Update installer will need to add to the trust store. 

So which host you use to access the site itself is a tradeoff between speed (Cloudflare edge cache) and security. WSUS can be configured to just always use the SHA1 host on XP/2000 because Cloudflare can’t cache it anyway.

**Comment by InternalLoss** [comment]
Sounds like a good idea :)

Do you have a Discord? it might be easier to communicate there than cluttering up GitHub Issues - I'm currently trying to test Win2k RTM and running into issues I'm sure would be easy to fix :P

**Comment by kirb** [comment]
Definitely, send me a friend request at @kirb#1809 (https://chariz.com/discord if you can't find me, I'll be at the top of the member list)

**Comment by sydia1103** [comment]
Any update on this @kirb? It would definitely be be nice to access the site over HTTPS on XP and 2000.

**Comment by kirb** [comment]
Not really any progress on the server side of things yet, but I just released 1.2 which checks whether your machine is capable of connecting to the site over HTTPS, and if so, configures the WSUS server and Start menu shortcut to connect via HTTPS.

**Comment by sydia1103** [comment]
Also, just letting you know, XP can't access the site at all over HTTPS even with SP3, due to your server not supporting the 3DES cipher. Cloudflare does support it, as shown [here](https://developers.cloudflare.com/ssl/reference/cipher-suites/recommendations/), so maybe enabling it would be a start in making your website HTTPS compatible on XP. You can always check what ciphers your website supports [here](https://www.ssllabs.com/ssltest/analyze.html?d=legacyupdate.net).

**Comment by InternalLoss** [comment]
Just to follow along with above, modern Nginx from Ubuntu's APT doesn't seem to come with the ciphers needed for XP anymore (even if you use Mozilla's old SSL config)

**Comment by kirb** [comment]
Thanks heaps for that @sydia1103, though edge certificates is a feature of the advanced certificate manager feature which is a paid addon. I’m also not sure it’ll work without SNI, because if I understand right the idea is you’re purchasing a cert for yourself alone, unlike Cloudflare’s standard cert provisioning where they use one cert with hundreds of domains on it. I’ll do more research and likely enquire with them to get more info on its legacy compatibility.

@InternalLoss Good to know, I’ll probably have to run nginx in a Docker container with a custom build of OpenSSL or an older Ubuntu LTS.
